NOTES-daneroo.md

```diff
@@ -0,0 +1,104 @@
+
+Another issue, although the `.lasDone` being captured on a successful run is useful, as photos are listed in EXIF date order. If older photos are added to the album they would not appear in subsequent runs. So it would be useful in general to rerun over the entire album (`-all`).
+
+## Headless
+
+If we don't need the authentication flow, (and we persist the auth creds in the profile `-dev`), Chrome can be brought up in `-headless` mode, which considerably reduces memory footprint (`>2Gb` to `<500Mb`), and marginally (23%) increases throughput for the listing experiment.
+
+## `listFromAlbum()`
+
+This is another experiment where the listing was entirely performed in the main album page (by scrolling incrementally). This is even faster. It would also allow iterating either forward or backward through the album.
+
+In a further experiment, I would like to use this process as a coordinating mechanism and perform the actual downloads in separate (potentially multiple) `tabs/contexts`.
+
+### Performance: An Argument for a periodic page reload and headless mode
+
+Notice how the latency grows quickly without page reload (ms per iteration): [66, 74, 89, 219, 350, 1008,...], and the cumulative rate drops below `4 items/s` after `6000` items.
+
+```bash
+$ time ./gphotos-cdp  -dev -n 6001 -list
+21:36:26 . Avg Latency (last 1000 @ 1000): Marginal: 66.43ms Cumulative: 66.43ms
+21:37:40 . Avg Latency (last 1000 @ 2000): Marginal: 73.99ms Cumulative: 70.21ms
+21:39:09 . Avg Latency (last 1000 @ 3000): Marginal: 88.97ms Cumulative: 76.46ms
+21:42:48 . Avg Latency (last 1000 @ 4000): Marginal: 218.75ms Cumulative: 112.03ms
+21:48:38 . Avg Latency (last 1000 @ 5000): Marginal: 350.23ms Cumulative: 159.67ms
+22:05:27 . Avg Latency (last 1000 @ 6000): Marginal: 1008.58ms Cumulative: 301.16ms
+22:05:27 Rate (6001): 3.32/s Avg Latency: 301.11ms
+```
+
+Whereas with page reloading every 1000 images, it is stable at `<60ms` with a sustained rate of `17 items/s`:
+
+```bash
+$ time ./gphotos-cdp  -dev  -vt -list -n 6001
+14:33:22 . Avg Latency (last 1000 @ 1000):  Marginal: 58.88ms Cumulative: 58.88ms
+14:34:21 . Avg Latency (last 1000 @ 2000):  Marginal: 57.60ms Cumulative: 58.24ms
+14:35:21 . Avg Latency (last 1000 @ 3000):  Marginal: 57.93ms Cumulative: 58.14ms
+14:36:17 . Avg Latency (last 1000 @ 4000):  Marginal: 54.93ms Cumulative: 57.33ms
+14:37:16 . Avg Latency (last 1000 @ 5000):  Marginal: 56.24ms Cumulative: 57.12ms
+14:38:16 . Avg Latency (last 1000 @ 6000):  Marginal: 58.73ms Cumulative: 57.38ms
+14:38:16 Rate (6001): 17.43/s Avg Latency: 57.38ms
+```
```

Comment:
tbh, I don't think we care that much about performance, because as I said before, we kinda want to mimic an end-user behaviour. So even a download rate as low as 4 photos/s is pretty high, compared to what a user would do manually.

Reply:
I understand. 
The use case I had in mind (which may be out of scope for this) is to support multiple profiles (users), and run periodically from CRON i.e. daily incremental, weekly full verification). In that case, speed can become more important, especially for accounts with huge numbers of photos.
I have started working on this as a separate effort (in node.js with puppeteer, and can report back when I'm done, to see if you are interested in any of that.

Reply:
sure